python/ray/experimental/channel/cpu_nccl_group.py

```diff
@@ -0,0 +1,169 @@
+                self.condition.notify_all()
+            else:
+                await self.condition.wait_for(lambda: self.num_actors_seen[op_id] == self.num_actors)
+
+            result = self.collective_data[op_id]
+
+            if self.num_actors_seen[op_id] == self.num_actors:
+                del self.collective_data[op_id]
+                del self.num_actors_seen[op_id]
+
+            return result
+
+    def _apply_op(self, op: ReduceOp, tensors: List[torch.Tensor]) -> torch.Tensor:
+        """Apply the specified reduction operation across a list of tensors."""
+        result = tensors[0].clone()
+        if op == ReduceOp.SUM:
+            for tensor in tensors[1:]:
+                result += tensor
+        elif op == ReduceOp.PRODUCT:
+            for tensor in tensors[1:]:
+                result *= tensor
+        elif op == ReduceOp.MAX:
+            for tensor in tensors[1:]:
+                result = torch.max(result, tensor)
+        elif op == ReduceOp.MIN:
+            for tensor in tensors[1:]:
+                result = torch.min(result, tensor)
+        elif op == ReduceOp.AVG:
+            result = sum(tensors) / len(tensors)
+        else:
+            raise ValueError(f"Operation {op} not supported")
+        return result
+
+
+class CPUNcclGroup(ray_channel.nccl_group._NcclGroup):
+    """
+    Mock the internal _NcclGroup to use a communicator actor instead of a NCCL group
+    for communication.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        """We use the op index to synchronize the sender and receiver at the
+        communicator actor."""
+        self.num_ops = defaultdict(int)
+        self.communicators = set()
+
+    def _ensure_communicator_exists(self, comm_key: str, is_p2p: bool = False):
+        """Ensure the necessary communicator actor is created."""
+        try:
+            ray.get_actor(name=comm_key)
+        except ValueError:
+            ray.remote(CPUCommunicator).options(name=comm_key).remote(2 if is_p2p else self._world_size)
+
+    def send(self, tensor: torch.Tensor, peer_rank: int):
+        """Send the tensor to the communicator actor."""
+        comm_key = f"communicator-{self.get_self_rank()}-{peer_rank}"
+        self._ensure_communicator_exists(comm_key, is_p2p=True)
+        communicator = ray.get_actor(name=comm_key)
+
+        self.communicators.add(communicator)
+        ray.get(communicator.wait_p2p.remote(self.num_ops[comm_key], tensor))
+        self.num_ops[comm_key] += 1
+
+    def recv(
+        self,
+        shape: Tuple[int],
+        dtype: torch.dtype,
+        peer_rank: int,
+        allocator: Optional[TorchTensorAllocator] = None,
+    ):
+        """Receive the tensor from the communicator actor."""
+        comm_key = f"communicator-{peer_rank}-{self.get_self_rank()}"
+        self._ensure_communicator_exists(comm_key, is_p2p=True)
+        communicator = ray.get_actor(name=comm_key)
```

Comment:
There may be two consecutive `ray.get_actor` calls. Maybe we can merge these few lines into `_ensure_communicator_exists` and have it return the communicator. In that case, maybe rename it to something like `_get_or_create_communicator`.

Reply:
Fixed